lib/keyscheme.js

```diff
@@ -0,0 +1,318 @@
+'use strict'; // eslint-disable-line strict
+
+/**
+ * Module in charge of picking new chunk/part keys and
+ * their endpoints, aggregating those 'sub keys' into a
+ * single 'rawKey', serializing and deserializing raw keys
+ *
+ * Version 1 scheme grammar:
+ * <genkey> := <version>#<topologyVersion>#<split>#<repPolicy># \
+*                  <objectKey>#<rand>#<location>[#<location>]
+ * <version> := Natural (so 0 or 1 to start)
+ * <topologyVersion> := Natural (so 0 or 1 to start) - if we use indirection
+ *                      instead of storing
+ * <split> := <nChunk>%<splitSize>
+ * <nChunk> := number of splitted parts (1 for non-splitted objects)
+ * <splitSize> := size of each splitted parts, except last one
+ *                 (see hyperdrive keys below)
+ * <repPolicy> := RSk+m or CPn (for n-1 copies)
+ * <objectKey> := parent S3 object key (or a prefix) - can contains anything but
+ *                section and sub-section separators
+ * <rand> := 64 bits random number (unicity inside 1 hyperdrive)
+ * <location>:= hyperdrive location (UUID, idx in table?, ip:port, ...)
+ *
+ * Actual, stored hyperdrive keys can be easily computed from <genkey>.
+ * <storedFragmentKey> := <objectKey>-<rand>-<startOffset>-<type>-<fragid>
+ * <objectKey> and <rand> are the ones defined above
+ * <type> := ‘d’ for data, ‘c’ for coding
+ * <fragid>:= index in main key fragment list
+ * <startOffset> := used for splits. All split chunks share the same prefix,
+ *                  storing the offset is used to easily have range queries
+ *                  and avoid storing them all in the main key.
+ */
+
+const assert = require('assert');
+const placement = require('./placement');
+const split = require('./split');
+
+/**
+ * Separator used to mark sections in rawKey
+ */
+const SECTION_SEPARATOR = '#';
+
+/**
+ * Separator used to mark sub sections in rawKey
+ */
+const SUBSECTION_SEPARATOR = ',';
+
+/**
+ * Separator used to structure the part/fragment key.
+ * This key is the one actually sent and stored in
+ * the hyperdrives
+ */
+const PART_KEY_SEPARATOR = '-';
+
+/**
+ * Keyscheme is versioned, to easily support changes.
+ * Each generated rawKey embeds the keygen version.
+ */
+const KEYSCHEME_VERSION = 1;
+
+/* TODO if we don't store directly endpoints of each chunk,
+ * then endpoints topology MUST be versioned as well.
+ */
+const TOPOLOGY_VERSION = 1;
+
+class KeySchemeDeserializeError extends Error {
+    constructor(message) {
+        super(message);
+        // Saving class name in the property of our custom error as a shortcut.
+        this.name = this.constructor.name;
+        // Capturing stack trace, excluding constructor call from it.
+        Error.captureStackTrace(this, this.constructor);
+    }
+}
+
+
+function range(n) {
```

Comment:
Hey, you're not using that one much, given that I've seen the construct `[...Array().keys()]` a few times in the code ;)
It could thus be a more global util :)

Reply:
Indeed! I actually forgot I defined it here XD

Reply:
Done